{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio.Seq import Seq\n",
    "from collections import defaultdict\n",
    "import re\n",
    "import logging\n",
    "from Bio.SeqUtils import GC, GC123\n",
    "from Bio.Alphabet import generic_dna\n",
    "from collections import Counter\n",
    "from Bio.Seq import MutableSeq\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "from scipy import mean as sc_m\n",
    "import itertools\n",
    "import pandas as pd\n",
    "import researchpy as rp\n",
    "from scipy import stats\n",
    "import seaborn as sns\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.formula.api import ols\n",
    "import matplotlib.pyplot as plt\n",
    "from statsmodels.stats import anova \n",
    "from matplotlib import cm \n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd\n",
    "from statsmodels.stats.libqsturng import psturng\n",
    "from tabulate import tabulate\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "logging.basicConfig(filename = \"avisos_cambios.log\", filemode = \"w\", \n",
    "                    format = \"%(name)s - %(levelname)s - %(asctime)s - %(message)s\",\n",
    "                    level = logging.INFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ruta a la carpeta en la que se encuentran almacenados los archivos genbank\n",
    "directorio = \"./GenomasHPV/\"\n",
    "list_files = []\n",
    "for root, directories, files in os.walk(directorio):\n",
    "    for name in files:\n",
    "        list_files.append(os.path.join(root, name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ['E1', 'E2', 'E4', 'E5', 'E6', 'E7', 'L2', 'L1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#procesar los archivos genbank para obtener la secuencia genómica completa, \n",
    "#el concatenado y los genes por separado\n",
    "\n",
    "def process_genbank(list_files):\n",
    "    \n",
    "    genome_genes = defaultdict(dict)\n",
    "    whole_genome = defaultdict(dict)\n",
    "    genome_genes_cat = defaultdict(dict)\n",
    "\n",
    "    for file in list_files:\n",
    "        with open(file, \"r\") as f_hpv:\n",
    "            data = SeqIO.parse(f_hpv, \"genbank\")\n",
    "        \n",
    "            virus_genus = f_hpv.name.split(\"/\")[-2]\n",
    "            virus_type = (re.findall(r'\\d+', f_hpv.name.split(\"/\")[-1])[0])\n",
    "        \n",
    "            whole_genome_seq = []\n",
    "            genome_genes_seq = []\n",
    "            genes_cat_seq = []\n",
    "        \n",
    "            for seqrecord in data:\n",
    "            \n",
    "                whole_genome_seq.append(SeqRecord(seqrecord.seq, id = seqrecord.id, description = seqrecord.description))\n",
    "    \n",
    "                genes_cat = \"\"\n",
    "            \n",
    "                gen_present = {\"E1\" : False, \"E2\" : False, \"E4\" : False, \"E5\" : False, \"E6\" : False, \"E7\" : False,\n",
    "                          \"L1\" : False, \"L2\" : False}\n",
    "            \n",
    "                if seqrecord.features:\n",
    "                    gene=\"\"\n",
    "            \n",
    "                    for feature in seqrecord.features:\n",
    "\n",
    "                        if feature.type == \"gene\":\n",
    "                        \n",
    "                            if feature.qualifiers == \"gene\":\n",
    "                                gene = feature.qualifiers[\"gene\"][0]\n",
    "                                gene = gene.upper()\n",
    "                            \n",
    "                        elif feature.type == \"CDS\":\n",
    "                            if \"gene\" in feature.qualifiers:\n",
    "                                if feature.qualifiers[\"gene\"][0].split()[0] in genes:\n",
    "                                    cds = feature.qualifiers[\"gene\"][0].split()[0]\n",
    "                                    cds = cds.upper()\n",
    "                                \n",
    "                                    if not gen_present[cds]:\n",
    "                                        \n",
    "                                        gen_present[cds] = True\n",
    "                                    \n",
    "                                        if gene and (cds != gene): #error\n",
    "                                            warn = seqrecord.description + \";\" + cds\n",
    "                                            logging.warning('%s raised a warning, gene and CDS are different', warn)\n",
    "                                    #ignoraré este CDS porque el feature tipo gene y el cds.gene no coinciden\n",
    "                                        else: \n",
    "                                            seq_cds = feature.extract(seqrecord.seq)\n",
    "                                            genome_genes_seq.append(SeqRecord(seq_cds, id = seqrecord.id, name = cds, \n",
    "                                                              description = seqrecord.description))\n",
    "                                        \n",
    "                                            if gene != \"E5\":\n",
    "                                                genes_cat = genes_cat + seq_cds\n",
    "                                    else:\n",
    "                                        warn = seqrecord.description + \";\" + cds\n",
    "                                        logging.warning('%s raised a warning, gene is duplicated', warn)\n",
    "                                \n",
    "                                        \n",
    "                            #algunos genes están anotados como ORF pero se corresponden con las mismas unidades,\n",
    "                            # por lo que si coinciden las tenemos en cuenta igual\n",
    "                                elif \"ORF\" in feature.qualifiers[\"gene\"][0].split():\n",
    "                                \n",
    "                                    name_orf = feature.qualifiers[\"gene\"][0].split()\n",
    "                                    name_orf.remove(\"ORF\")\n",
    "                                \n",
    "                                    name_orf = name_orf[0].upper()\n",
    "                                \n",
    "                                    warn = seqrecord.description + \";\" + name_orf\n",
    "                                    logging.warning('%s raised a warning, gene name included ORF', warn)\n",
    "                                \n",
    "                                    if name_orf in genes:\n",
    "                                        cds = name_orf\n",
    "                                    \n",
    "                                        if not gen_present[cds]:\n",
    "                                    \n",
    "                                            gen_present[cds] = True\n",
    "\n",
    "                                            if gene and (cds != gene): #error\n",
    "                                                warn = seqrecord.description + \";\" + cds\n",
    "                                                logging.warning('%s raised a warning, gene and CDS are different', warn)\n",
    "                                            else: \n",
    "                                                seq_cds = feature.extract(seqrecord.seq)\n",
    "                                                genome_genes_seq.append(SeqRecord(seq_cds, id = seqrecord.id, name = cds, \n",
    "                                                              description = seqrecord.description))\n",
    "                                                if gene != \"E5\":\n",
    "                                                    genes_cat = genes_cat + seq_cds\n",
    "                                        else:\n",
    "                                            warn = seqrecord.description + \";\" + cds\n",
    "                                            logging.warning('%s raised a warning, gene is duplicated', warn)\n",
    "                                \n",
    "                                    else:\n",
    "                                        warn = seqrecord.description + \";\" + name_orf\n",
    "                                        logging.warning('%s raised a warning, gene name not in genes', warn)\n",
    "                            \n",
    "                            # Si el nombre del gen no está en nuestra lista y tampoco presenta ORF hay un \n",
    "                            # warning y lo ignoramos \n",
    "                                else:\n",
    "                                    warn = seqrecord.description + \";\" + feature.qualifiers[\"gene\"][0].split()[0]\n",
    "                                    logging.warning('%s raised a warning, gene name not in genes', warn)\n",
    "                       \n",
    "                        #si no hay cualificador gene en CDS miro si hay product\n",
    "                            elif \"product\" in feature.qualifiers: \n",
    "                                if feature.qualifiers[\"product\"][0].split()[0] in genes:\n",
    "                                    cds = feature.qualifiers[\"product\"][0].split()[0]\n",
    "                                    cds = cds.upper()\n",
    "                                \n",
    "                                    if not gen_present[cds]:\n",
    "                                        \n",
    "                                        gen_present[cds] = True\n",
    "    \n",
    "                                        if gene and (cds != gene): #error\n",
    "                                            warn = seqrecord.description + \";\" + cds\n",
    "                                            logging.warning('%s raised a warning, gene and CDS are different', warn)\n",
    "                                        else: \n",
    "                                            seq_cds = feature.extract(seqrecord.seq)\n",
    "                                            genome_genes_seq.append(SeqRecord(seq_cds, id = seqrecord.id, name = cds, \n",
    "                                                              description = seqrecord.description))\n",
    "                                        \n",
    "                                            if gene != \"E5\":\n",
    "                                                genes_cat = genes_cat + seq_cds\n",
    "                                    else:\n",
    "                                        warn = seqrecord.description + \";\" + cds\n",
    "                                        logging.warning('%s raised a warning, gene is duplicated', warn)\n",
    "                                \n",
    "                                                \n",
    "                                elif \"ORF\" in feature.qualifiers[\"product\"][0].split():\n",
    "                                \n",
    "                                    name_orf = feature.qualifiers[\"product\"][0].split()\n",
    "                                    name_orf.remove(\"ORF\")\n",
    "                                \n",
    "                                    name_orf = name_orf[0].upper()\n",
    "                                \n",
    "                                    warn = seqrecord.description + \";\" + name_orf\n",
    "                                    logging.warning('%s raised a warning, product name included ORF', warn)\n",
    "                                \n",
    "                                    if name_orf in genes:\n",
    "                                        cds = name_orf\n",
    "                                    \n",
    "                                        if not gen_present[cds]:\n",
    "                                    \n",
    "                                            gen_present[cds] = True\n",
    "\n",
    "                                            if gene and (cds != gene): #error\n",
    "                                                warn = seqrecord.description + \";\" + cds\n",
    "                                                logging.warning('%s raised a warning, gene and CDS are different', warn)\n",
    "                                        \n",
    "                                            else: \n",
    "                                                seq_cds = feature.extract(seqrecord.seq)\n",
    "                                                genome_genes_seq.append(SeqRecord(seq_cds, id = seqrecord.id, name = cds, \n",
    "                                                              description = seqrecord.description))\n",
    "                                                if gene != \"E5\":\n",
    "                                                    genes_cat = genes_cat + seq_cds\n",
    "                                        else:\n",
    "                                            warn = seqrecord.description + \";\" + cds\n",
    "                                            logging.warning('%s raised a warning, gene is duplicated', warn)\n",
    "                                \n",
    "                                    else:\n",
    "                                        warn = seqrecord.description + \";\" + name_orf\n",
    "                                        logging.warning('%s raised a warning, gene name not in genes', warn)\n",
    "                            \n",
    "                if len(genes_cat) > 0:\n",
    "                    genes_cat_seq.append(SeqRecord(genes_cat, id = seqrecord.id, description = seqrecord.description))\n",
    "        \n",
    "            whole_genome[virus_genus][virus_type] = whole_genome_seq\n",
    "        \n",
    "            if len(genes_cat_seq) > 0:\n",
    "                genome_genes_cat[virus_genus][virus_type] = genes_cat_seq\n",
    "                genome_genes[virus_genus][virus_type] = genome_genes_seq\n",
    "        \n",
    "        \n",
    "\n",
    "    return whole_genome, genome_genes_cat, genome_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#obtener valores GC \n",
    "def gc123_genomes(genomes_dict):\n",
    "    \n",
    "    gc123_output_list = defaultdict(dict)\n",
    "    \n",
    "    for virus_genus in genomes_dict.keys():\n",
    "        \n",
    "        for virus_type in genomes_dict[virus_genus].keys():\n",
    "        \n",
    "            gc123_values = []\n",
    "        \n",
    "            for record in genomes_dict[virus_genus][virus_type]:\n",
    "                \n",
    "                gc123 = list(GC123(record.seq))\n",
    "                gc12 = np.mean([gc123[1],gc123[2]])\n",
    "                gc123.append(gc12)\n",
    "            \n",
    "                gc123_values.append(gc123)\n",
    "\n",
    "            gc123_output_list[virus_genus][virus_type] = gc123_values\n",
    "    \n",
    "    return gc123_output_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_genomes_gc_original(gc123_list):\n",
    "    \n",
    "    df_final = pd.DataFrame()\n",
    "    \n",
    "    for virus_genus in gc123_list.keys():\n",
    "        \n",
    "        for virus_type in gc123_list[virus_genus].keys():\n",
    "        \n",
    "            df_gc = pd.DataFrame.from_dict(gc123_list[virus_genus][virus_type])\n",
    "            \n",
    "            col_gen = [virus_genus] * len(gc123_list[virus_genus][virus_type])\n",
    "            col_type = [virus_type] * len(gc123_list[virus_genus][virus_type])\n",
    "\n",
    "            df_gc[\"Genero\"] = col_gen\n",
    "            df_gc[\"Tipo\"] = col_type\n",
    "            \n",
    "            cols = df_gc.columns.tolist()\n",
    "            cols = [cols[5], cols[6], cols[0], cols[4], cols[3]]\n",
    "            df_gc = df_gc[cols]\n",
    "            df_gc.rename(columns = {0: \"Total_GC\", 4:\"GC12\", 3:\"GC3\"}, inplace = True)\n",
    "            \n",
    "            \n",
    "            \n",
    "            df_final = pd.concat([df_final, df_gc], axis = 0)\n",
    "        \n",
    "    return df_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#eliminar codones seleccionados de la secuencia de los genes \n",
    "\n",
    "def gc3s_genes(genome_genes):\n",
    "    genomes_dict_by_type_and_gen_gc3s = defaultdict(dict)\n",
    "\n",
    "    for virus_genus in genome_genes.keys():\n",
    "    \n",
    "        for virus_type in genome_genes[virus_genus].keys():\n",
    "    \n",
    "            genomes_dict_by_gen_gc3s = defaultdict(list)\n",
    "        \n",
    "            for record in genome_genes[virus_genus][virus_type]:\n",
    "    \n",
    "                sequ = Seq(str(record.seq))\n",
    "                codons = [sequ[i:i+3] for i in range(0, len(sequ), 3)]\n",
    "                delete = [\"ATG\", \"TGA\", \"TAG\", \"TAA\", \"TGG\"]\n",
    "    \n",
    "                sequence = \"\"\n",
    "        \n",
    "                for triplete in codons:\n",
    "                    if triplete not in delete:\n",
    "                        sequence = sequence + triplete\n",
    "\n",
    "                try: \n",
    "                    len(sequence) % 3 != 0\n",
    "                \n",
    "                except:\n",
    "                    err = virus_genus + \";\" + virus_type \n",
    "                    logging.error(\"%s raised an error, it cannot be divided by 3\")\n",
    "            \n",
    "            \n",
    "                genomes_dict_by_gen_gc3s[record.name].append(sequence)\n",
    "            \n",
    "            genomes_dict_by_type_and_gen_gc3s[virus_genus][virus_type] = genomes_dict_by_gen_gc3s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtener los valores GC en el caso de los genes\n",
    "\n",
    "def gc_genes(genomes_dict_by_type_and_gen_gc3s):\n",
    "    gc123_by_type_and_gene_list_todas = defaultdict(dict)\n",
    "\n",
    "    for virus_genus in genomes_dict_by_type_and_gen_gc3s.keys():\n",
    "    \n",
    "        for virus_type in genomes_dict_by_type_and_gen_gc3s[virus_genus]:\n",
    "        \n",
    "            medias_by_gene = {}\n",
    "            desv_by_gene = {}\n",
    "        \n",
    "            available_genes = genomes_dict_by_type_and_gen_gc3s[virus_genus][virus_type].keys()\n",
    "        \n",
    "            gc123_by_gen_dict = defaultdict(list)\n",
    "            \n",
    "            for gene in available_genes:\n",
    "            \n",
    "                for sequence in genomes_dict_by_type_and_gen_gc3s[virus_genus][virus_type][gene]:\n",
    "                \n",
    "                    gc123 = list(GC123(sequence))\n",
    "                    gc12 = np.mean([gc123[1],gc123[2]])\n",
    "                    gc123.append(gc12)\n",
    "                    gc123_by_gen_dict[gene].append(gc123)\n",
    "                \n",
    "                \n",
    "                gc123_by_gen_gc3s_array = np.asarray(gc123_by_gen_dict[gene])\n",
    "            \n",
    "            gc123_by_type_and_gene_list_todas[virus_genus][virus_type] = gc123_by_gen_dict\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_genomes_genes_gc_original(gc123_genes_list):\n",
    "    \n",
    "    df_final = pd.DataFrame()\n",
    "    \n",
    "    for virus_genus in gc123_genes_list.keys():\n",
    "        \n",
    "        for virus_type in gc123_genes_list[virus_genus].keys():\n",
    "            \n",
    "            for gene in gc123_genes_list[virus_genus][virus_type].keys():\n",
    "                \n",
    "                df_gc = pd.DataFrame.from_dict(gc123_genes_list[virus_genus][virus_type][gene])\n",
    "                \n",
    "                \n",
    "                col_genero = [virus_genus] * len(gc123_genes_list[virus_genus][virus_type][gene])\n",
    "                col_type = [virus_type] * len(gc123_genes_list[virus_genus][virus_type][gene])\n",
    "                col_gen = [gene] * len(gc123_genes_list[virus_genus][virus_type][gene])\n",
    "                \n",
    "                \n",
    "                df_gc[\"Genero\"] = col_genero\n",
    "                df_gc[\"Tipo\"] = col_type\n",
    "                df_gc[\"Gen\"] = col_gen\n",
    "                \n",
    "                cols = df_gc.columns.tolist()\n",
    "                cols = [cols[5], cols[6], cols[7], cols[0], cols[4], cols[3]]\n",
    "                df_gc = df_gc[cols]\n",
    "                df_gc.rename(columns = {0: \"Total_GC\", 4:\"GC12\", 3:\"GC3\"}, inplace = True)\n",
    "            \n",
    "                df_final = pd.concat([df_final, df_gc], axis = 0)\n",
    "        \n",
    "    return df_final"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dinucleótidos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#obtener la lista de dinucleótidos\n",
    "nucl = [\"A\", \"C\", \"T\", \"G\"]\n",
    "dinucl = [p for p in itertools.product(nucl, repeat = 2)]\n",
    "dinucleotidos = []\n",
    "\n",
    "for pair in dinucl:\n",
    "    dinucleotidos.append(pair[0]+pair[1])\n",
    "#print(dinucleotidos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determinar que nucleótidos hay que tener en cuenta para el cálculo de\n",
    "#frecuencias en función de los que componen el dinucleótido\n",
    "\n",
    "def value_dinucleotidos(pair, C_total, G_total, T_total, A_total, din_dict):\n",
    "\n",
    "        din_dict[pair]\n",
    "        \n",
    "        x = {}\n",
    "        y = {}\n",
    "        if \"A\" in pair:\n",
    "            x = A_total\n",
    "            if pair == \"AA\":\n",
    "                y = A_total\n",
    "            \n",
    "        if \"T\" in pair:\n",
    "            if len(x.keys()) == 0:\n",
    "                x = T_total\n",
    "                if pair == \"TT\":\n",
    "                    y = T_total\n",
    "            else: \n",
    "                y = T_total\n",
    "     \n",
    "        if \"C\" in pair:\n",
    "            if len(x.keys()) == 0:\n",
    "                x = C_total\n",
    "                if pair == \"CC\":\n",
    "                    y = C_total\n",
    "            else: \n",
    "                y = C_total\n",
    " \n",
    "        if \"G\" in pair:\n",
    "            if len(x.keys()) == 0:\n",
    "                x = G_total\n",
    "                if pair == \"GG\":\n",
    "                    y = G_total\n",
    "            else: \n",
    "                y = G_total\n",
    "                \n",
    "        return(x, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cálculo de la frecuencia de dinucleótidos cuando no se tiene en cuenta la posición\n",
    "def freq_dinucleotidos(genome):\n",
    "    \n",
    "    C_total = defaultdict(dict)\n",
    "    G_total = defaultdict(dict)\n",
    "\n",
    "    T_total = defaultdict(dict)\n",
    "    A_total = defaultdict(dict)\n",
    "\n",
    "    dinucleotidos_por_tipo = defaultdict(dict)\n",
    "\n",
    "    for virus_genus in genome.keys():\n",
    "    \n",
    "        for virus_type in genome[virus_genus].keys():\n",
    "            \n",
    "            din_dict = defaultdict(list)\n",
    "        \n",
    "            for pair in dinucleotidos:\n",
    "                din_dict[pair]\n",
    "        \n",
    "            C_total_list = [] \n",
    "            G_total_list = [] \n",
    "            T_total_list = [] \n",
    "            A_total_list = [] \n",
    "        \n",
    "            for record in genome[virus_genus][virus_type]:\n",
    "            \n",
    "                sequence = Seq(str(record.seq))\n",
    "                counter = Counter(sequence)\n",
    "    \n",
    "    \n",
    "                C_total_list.append((counter[\"C\"])/len(sequence))\n",
    "                G_total_list.append((counter[\"G\"])/len(sequence))\n",
    "                T_total_list.append((counter[\"T\"])/len(sequence))\n",
    "                A_total_list.append((counter[\"A\"])/len(sequence))\n",
    "    \n",
    "        \n",
    "    \n",
    "                for pair in dinucleotidos:\n",
    "            \n",
    "                    frequency = freq_overlapping(sequence, pair)\n",
    "\n",
    "                    din_dict[pair].append(frequency)\n",
    "        \n",
    "            C_total[virus_genus][virus_type] = C_total_list\n",
    "            G_total[virus_genus][virus_type] = G_total_list\n",
    "            T_total[virus_genus][virus_type] = T_total_list\n",
    "            A_total[virus_genus][virus_type] = A_total_list\n",
    "        \n",
    "        \n",
    "            dinucleotidos_por_tipo[virus_genus][virus_type] = din_dict\n",
    "            \n",
    "    return dinucleotidos_por_tipo, C_total, G_total, T_total, A_total\n",
    "\n",
    "\n",
    "\n",
    "#cálculo de la frecuencia de dicucleótidos teniendo en cuenta la posición\n",
    "def freq_dinucleotidos_123_genome(genome):\n",
    "    \n",
    "    dinucleotidos_12 = defaultdict(dict)\n",
    "    dinucleotidos_23 = defaultdict(dict)\n",
    "    \n",
    "    for virus_genus in genome.keys():\n",
    "                \n",
    "        for virus_type in genome[virus_genus].keys():\n",
    "            \n",
    "            din_dict_12 = defaultdict(list)\n",
    "            din_dict_23 = defaultdict(list)\n",
    "\n",
    "            for sequence in genome[virus_genus][virus_type]:\n",
    "                \n",
    "                sequ = Seq(str(sequence.seq))\n",
    "                \n",
    "                codons = [sequ[i:i+3] for i in range(0, len(sequ), 3)]\n",
    "                codons_12 = [codons[i][0:2] for i in range (len(codons))]\n",
    "                codons_23 = [codons[i][1:3] for i in range (len(codons))]\n",
    "    \n",
    "                count_12 = Counter(codons_12)\n",
    "                count_23 = Counter(codons_23)\n",
    "    \n",
    "                for pair in dinucleotidos:\n",
    "            \n",
    "                    freq_12 = count_12[pair]/len(codons)\n",
    "                    din_dict_12[pair].append(freq_12)\n",
    "                    \n",
    "                    freq_23 = count_23[pair]/len(codons)\n",
    "                    din_dict_23[pair].append(freq_23)\n",
    "            \n",
    "            dinucleotidos_12[virus_genus][virus_type] = din_dict_12\n",
    "            dinucleotidos_23[virus_genus][virus_type] = din_dict_23\n",
    "            \n",
    "    \n",
    "    return dinucleotidos_12, dinucleotidos_23\n",
    "\n",
    "\n",
    "#cálculo de la frecuencia observados/esperados\n",
    "def obs_esp_genoma(frequency_dinucleotidos, C_total, G_total, T_total, A_total):\n",
    "\n",
    "    \n",
    "    obs_esp_dinucl_por_tipo = defaultdict(dict)\n",
    "\n",
    "    for virus_genus in frequency_dinucleotidos.keys():\n",
    "    \n",
    "        for virus_type in frequency_dinucleotidos[virus_genus].keys():\n",
    "        \n",
    "            din_dict = defaultdict(list)\n",
    "        \n",
    "            for pair in dinucleotidos:\n",
    "            \n",
    "                x, y = value_dinucleotidos(pair, C_total, G_total, T_total, A_total, din_dict)\n",
    "              \n",
    "                for value in range(len(frequency_dinucleotidos[virus_genus][virus_type][pair])):\n",
    "\n",
    "                    obs_esp = frequency_dinucleotidos[virus_genus][virus_type][pair][value] / \\\n",
    "                                            (x[virus_genus][virus_type][value] * y[virus_genus][virus_type][value])\n",
    "                    din_dict[pair].append(obs_esp)\n",
    "            \n",
    "            obs_esp_dinucl_por_tipo[virus_genus][virus_type] = din_dict\n",
    "    return obs_esp_dinucl_por_tipo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_genomes_dinucleotidos_original(obs_esp):\n",
    "    \n",
    "    df_final = pd.DataFrame()\n",
    "    \n",
    "    for virus_genus in obs_esp.keys():\n",
    "        \n",
    "        for virus_type in obs_esp[virus_genus].keys():\n",
    "            \n",
    "            df_din = pd.DataFrame()\n",
    "            \n",
    "            for pair in obs_esp[virus_genus][virus_type].keys():\n",
    "            \n",
    "                df_din_pair = pd.DataFrame.from_dict(obs_esp[virus_genus][virus_type][pair])\n",
    "            \n",
    "                df_din_pair.rename(columns = {0: pair}, inplace = True)\n",
    "                #print(df_din_pair)\n",
    "                \n",
    "                df_din = pd.concat([df_din, df_din_pair], axis = 1)\n",
    "                \n",
    "            col_gen = [virus_genus] * len(df_din)\n",
    "            col_type = [virus_type] * len(df_din)\n",
    "            \n",
    "            df_din[\"Genero\"] = col_gen\n",
    "            df_din[\"Tipo\"] = col_type\n",
    "                \n",
    "            df_final = pd.concat([df_final, df_din], axis = 0)\n",
    "        \n",
    "    return df_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cálculo de la frecuencia de los dinucleótidos en los genes, \n",
    "#sin tener en cuenta la posición\n",
    "\n",
    "def freq_dinucleotidos_genes(genome_by_genes):\n",
    "    \n",
    "    C_total = defaultdict(dict)\n",
    "    G_total = defaultdict(dict)\n",
    "\n",
    "    T_total = defaultdict(dict)\n",
    "    A_total = defaultdict(dict)\n",
    "\n",
    "    dinucleotidos_por_tipo = defaultdict(dict)\n",
    "\n",
    "    for virus_genus in genome_by_genes.keys():\n",
    "    \n",
    "        for virus_type in genome_by_genes[virus_genus].keys():\n",
    "            \n",
    "            din_dict_genes = defaultdict(dict)\n",
    "            C_total_list = defaultdict(list) \n",
    "            G_total_list = defaultdict(list)\n",
    "            T_total_list = defaultdict(list)\n",
    "            A_total_list = defaultdict(list) \n",
    "            \n",
    "            for gene in genome_by_genes[virus_genus][virus_type].keys():\n",
    "                \n",
    "                din_dict = defaultdict(list)\n",
    "        \n",
    "                for pair in dinucleotidos:\n",
    "                    din_dict[pair]\n",
    "        \n",
    "                \n",
    "                for record in genome_by_genes[virus_genus][virus_type][gene]:\n",
    "                    \n",
    "                    counter = Counter(record)\n",
    "    \n",
    "        \n",
    "                    C_total_list[gene].append((counter[\"C\"])/len(record))\n",
    "                    G_total_list[gene].append((counter[\"G\"])/len(record))\n",
    "                    T_total_list[gene].append((counter[\"T\"])/len(record))\n",
    "                    A_total_list[gene].append((counter[\"A\"])/len(record))\n",
    "    \n",
    "                    for pair in dinucleotidos:\n",
    "            \n",
    "                        frequency = freq_overlapping(record, pair)\n",
    "\n",
    "                        din_dict[pair].append(frequency)\n",
    "        \n",
    "                \n",
    "                din_dict_genes[gene] = din_dict\n",
    "            C_total[virus_genus][virus_type] = C_total_list\n",
    "            G_total[virus_genus][virus_type] = G_total_list\n",
    "            T_total[virus_genus][virus_type] = T_total_list\n",
    "            A_total[virus_genus][virus_type] = A_total_list\n",
    "        \n",
    "            dinucleotidos_por_tipo[virus_genus][virus_type] = din_dict_genes\n",
    "            \n",
    "    return dinucleotidos_por_tipo, C_total, G_total, T_total, A_total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cálculo de la frecuencia de los dinucleótidos en los genes,\n",
    "#teniendo en cuenta la posición\n",
    "\n",
    "def freq_dinucleotidos_123_genes(genome):\n",
    "    \n",
    "    dinucleotidos_12 = defaultdict(dict)\n",
    "    dinucleotidos_23 = defaultdict(dict)\n",
    "    \n",
    "    for virus_genus in genome.keys():\n",
    "                \n",
    "        for virus_type in genome[virus_genus].keys():\n",
    "            \n",
    "            din_dict_genes_12 = defaultdict(dict)\n",
    "            din_dict_genes_23 = defaultdict(dict)\n",
    "            \n",
    "            for gen in genome[virus_genus][virus_type].keys():\n",
    "                \n",
    "                din_dict_12 = defaultdict(list)\n",
    "                din_dict_23 = defaultdict(list)\n",
    "            \n",
    "                for sequence in genome[virus_genus][virus_type][gen]:\n",
    "    \n",
    "                    codons = [sequence[i:i+3] for i in range(0, len(sequence), 3)]\n",
    "                    codons_12 = [codons[i][0:2] for i in range (len(codons))]\n",
    "                    codons_23 = [codons[i][1:3] for i in range (len(codons))]\n",
    "    \n",
    "                    count_12 = Counter(codons_12)\n",
    "                    count_23 = Counter(codons_23)\n",
    "    \n",
    "                    for pair in dinucleotidos:\n",
    "                        freq_12 = count_12[pair]/len(codons)\n",
    "                        din_dict_12[pair].append(freq_12)\n",
    "                        freq_23 = count_23[pair]/len(codons)\n",
    "                        din_dict_23[pair].append(freq_23)\n",
    "                \n",
    "                din_dict_genes_12[gen] = din_dict_12\n",
    "                din_dict_genes_23[gen] = din_dict_23\n",
    "                \n",
    "            dinucleotidos_12[virus_genus][virus_type] = din_dict_genes_12\n",
    "            dinucleotidos_23[virus_genus][virus_type] = din_dict_genes_23\n",
    "    \n",
    "    return dinucleotidos_12, dinucleotidos_23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cálculo de la frecuencia observados/esperados en los genes\n",
    "\n",
    "def obs_esp_genes(frequency_dinucleotidos, C_total, G_total, T_total, A_total):\n",
    "\n",
    "    \n",
    "    obs_esp_dinucl_por_tipo = defaultdict(dict)\n",
    "\n",
    "    for virus_genus in frequency_dinucleotidos.keys():\n",
    "    \n",
    "        for virus_type in frequency_dinucleotidos[virus_genus].keys():\n",
    "            \n",
    "            din_dict_genes = defaultdict(dict)\n",
    "        \n",
    "            for gene in frequency_dinucleotidos[virus_genus][virus_type].keys():\n",
    "                \n",
    "                din_dict = defaultdict(list)\n",
    "        \n",
    "                for pair in dinucleotidos:\n",
    "            \n",
    "                    x, y = value_dinucleotidos(pair, C_total, G_total, T_total, A_total, din_dict)\n",
    "                    \n",
    "                    for value in range(len(frequency_dinucleotidos[virus_genus][virus_type][gene][pair])):\n",
    "\n",
    "                        obs_esp = frequency_dinucleotidos[virus_genus][virus_type][gene][pair][value] / \\\n",
    "                                                (x[virus_genus][virus_type][gene][value] * \n",
    "                                                 y[virus_genus][virus_type][gene][value])\n",
    "                        din_dict[pair].append(obs_esp)\n",
    "                \n",
    "                din_dict_genes[gene] = din_dict\n",
    "            \n",
    "            obs_esp_dinucl_por_tipo[virus_genus][virus_type] = din_dict_genes\n",
    "                \n",
    "    return obs_esp_dinucl_por_tipo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_genomes_genes_gc_original(obs_esp_gen):\n",
    "    \n",
    "    df_final = pd.DataFrame()\n",
    "    \n",
    "    for virus_genus in obs_esp_gen.keys():\n",
    "        \n",
    "        for virus_type in obs_esp_gen[virus_genus].keys():\n",
    "            \n",
    "            for gene in obs_esp_gen[virus_genus][virus_type].keys():\n",
    "                \n",
    "                df_din = pd.DataFrame()\n",
    "                \n",
    "                for pair in obs_esp_gen[virus_genus][virus_type][gene].keys():\n",
    "                    \n",
    "                    df_din_pair = pd.DataFrame.from_dict(obs_esp_gen[virus_genus][virus_type][gene])\n",
    "            \n",
    "                    df_din_pair.rename(columns = {0: pair}, inplace = True)\n",
    "                \n",
    "                    df_din = pd.concat([df_din, df_din_pair], axis = 1)\n",
    "                \n",
    "                col_genero = [virus_genus] * len(df_din)\n",
    "                col_type = [virus_type] * len(df_din)\n",
    "                col_gen = [gene] * len(df_din)\n",
    "                \n",
    "                df_din[\"Genero\"] = col_genero\n",
    "                df_din[\"Tipo\"] = col_type\n",
    "                df_din[\"Gen\"] = col_gen\n",
    "                \n",
    "                df_final = pd.concat([df_final, df_din], axis = 0)\n",
    "        \n",
    "    return df_final"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
